### DISCUSSION TITLE
Settings window can not be opened in a space with a fullscreen app

### DISCUSSION THREAD

**Discussion by jfid** [orchetect/SettingsAccess#8] [Q&A] [answered]
### Bug Description, Steps to Reproduce, Crash Logs, Screenshots, etc.

I have developed a small sample app using packages MenuBarExtraAccess (of type .menuBarExtraStyle(.window)) and SettingsAccess. Basically, when clicking the status bar icon a small window appears with a button, that will open a Settings window.

<img width="544" alt="Screenshot 2023-10-23 at 23 43 32" src="https://github.com/orchetect/SettingsAccess/assets/203650/815b0040-d099-4553-b2f8-1daf90a488b8">

When I run the sample app in the Desktop with apps in non-fullscreen, everything works OK. But if I run the app from a Desktop space being used by an app in fullscreen mode, the main window appears correctly, and when I click the button to open the Settings, the Settings window does **NOT** appear in the Desktop space of the fullscreen app. It appears **SILENTLY** in the main Desktop. But the macOS focus keeps in the fullscreen app.

A couple of interesting comments:

- For displaying the SettingsView, I've tried both ways explained in the description (opening settings programatically, and opening settings using a MenuBarExtra. No luck with any of these options.

- I've run MacControlCenterUI demo and it shows the same problem.

Possible solutions:

Maybe the Settings window could be open with the window option set to floating. This would open it in the fullscreen Desktop:

`window.level = .floating`

Or make the Settings window as key. This would still show the Settings window in the main Desktop, but the macOS focus window will be set to it, so macOS will shown immediately the main Desktop with the Settings window in focus:

`window.makeKey()`

Would any of these solutions mitigate this problem?

**Comment by orchetect** [comment]
What macOS version is it?

Have you tried reproducing this in a new macOS app project with vanilla SwiftUI using just a window-based `MenuBarExtra` and `SettingsLink`? It would be useful to know if this is a SwiftUI issue or an issue related to the packages.

**Comment by jfid** [comment]
I'm in macOS 14.0, Xcode target set to 14.0.

Yes, a brand new project using SwiftUI, window-based MenuBarExtra and SettingsLink.

It could be a SwiftUI issue. If this is the case, is it still possible to capture the Settings window using `NSapp.windows` in some part of your package code, and try any of the solutions proposed? Something like:

```
for window in NSApp.windows {
    if let windowId = window.identifier?.rawValue {
        if windowId.contains("Settings") {
            window.level = .floating  // solution 1
            window.becomeKey()  // solution 2
        }
    }
}
```

**Comment by orchetect** [comment]
SettingsAccess is really just designed as a wrapper for Apple's `SettingsLink`. If there's a quirk or bug with Apple's implementation it's best to report that with a radar of course.

I imagine you could shoehorn that code into the post-action closure of the custom `SettingsLink` provided by the package. But it's brittle and may not cover all bases. It may be best left to the developer to solve this particular issue in that case.

**Comment by jfid** [comment]
Still fighting with it.

I've tried multiple code alternatives in the post-action closure, but none worked out. I had some partial success, for example I was able to hide the minimise and maximise buttons of the Settings window, but no luck when trying to make it `.floating` or set the focus calling `makeKey()` or `becomeKey()`. No worries, let's forget about this.

Then I came back to the original macControlCenterUI Demo app, and I realised that it **does** what I wanted. If I open the Settings from a fullscreen desktop, it **leaves** the fullscreen desktop and opens it in the main desktop setting the focus on it.

But creating a dummy sample project from scratch using both packages and doing same steps as macControlCenterUI Demo app does, I got different results. If I open the Settings from a fullscreen desktop, it **doesn't** leave the fullscreen desktop and opens it in the main desktop, without setting the focus on it.

Definitively I'm missing something but I can't find it.

```swift
import SwiftUI
import MenuBarExtraAccess
import SettingsAccess

@main
struct MenuBarPackagesApp: App {
    @State var isMenuPresented: Bool = false
    
    var body: some Scene {
        Settings {
            SettingsView()
        }

        MenuBarExtra("MyApp Menu", systemImage: "folder") {
            MyMenu(isMenuPresented: $isMenuPresented)
                .openSettingsAccess()
        }
        .menuBarExtraStyle(.window)
        .menuBarExtraAccess(isPresented: $isMenuPresented) { statusItem in
            // access status item or store it in a @State var
        }                
    }
}

struct SettingsView: View {
    var body: some View {
        Text("Hello")
            .frame(width: 300, height: 300)
    }
}

struct MyMenu: View {
    @Environment(\.openSettings) var openSettings
    @Binding var isMenuPresented: Bool

    var body: some View {
        Button("Open settings") {
            isMenuPresented = false
            try? openSettings()
        }
        .frame(width: 200, height: 200)
    }
}
```

**Comment by jfid** [comment, marked as answer]
Oh man, I found it.

I wasn't setting the Target property `"Application is agent (UIElement)"=YES`. For some reason that had a weird effect.

Doing that in fact opens the Settings window in the current fullscreen desktop. Whatever, that's good to me!

Let me thank you for your excellent work on these two packages, Steffan.

**Reply by orchetect** [reply]
Oh good! It's nice to find a simple resolution even if it takes time.
